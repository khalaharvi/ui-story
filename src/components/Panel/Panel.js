import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import _get from 'lodash.get';

import { Plus as PlusIcon } from 'icons';
import { cleanButton } from 'styles';

const Panel = styled('div')(({ theme }) => ({
  padding: '20px 5px',
  borderTop: `1px solid ${_get(theme, 'colors.secondary.taupe')}`,
  color: _get(theme, 'text.primary'),

  ':last-child': {
    borderBottom: `1px solid ${_get(theme, 'colors.secondary.taupe')}`
  }
}));

const PanelHeader = styled('div')(({ theme }) => ({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  ..._get(theme, 'typography.pSmall', {}),
  cursor: 'pointer',

  span: {
    flex: '1 1 0%'
  }
}));

const PanelToggle = styled('button')(({ theme }) => ({
  ...cleanButton,
  color: _get(theme, 'text.primary'),

  svg: {
    display: 'block',
    width: 16,
    height: 16,
    fill: 'currentColor'
  }
}));

const PanelBody = styled('div')(({ visible }) => ({
  display: visible ? 'block' : 'none',
  marginTop: 10
}));

// Adapted from: https://github.com/algolia/react-instantsearch/blob/v5.0.1/packages/react-instantsearch/src/components/Panel.js
class PanelComponent extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    header: PropTypes.string.isRequired
  };

  static childContextTypes = {
    setCanRefine: PropTypes.func.isRequired
  };

  state = {
    canRefine: true,
    toggled: false
  };

  getChildContext() {
    return {
      setCanRefine: this.setCanRefine
    };
  }

  setCanRefine = (nextCanRefine) => {
    this.setState({ canRefine: nextCanRefine });
  };

  toggle = () => {
    const { toggled } = this.state;

    this.setState({
      toggled: !toggled
    });
  }

  render() {
    const {
      children,
      header
    } = this.props;
    const { canRefine, toggled } = this.state;

    return (
      <Panel hidden={!canRefine}>
        <PanelHeader onClick={this.toggle}>
          <span>{header}</span>
          <PanelToggle type="button" aria-label={`Toggle ${header} panel`}>
            <PlusIcon rotate={toggled} />
          </PanelToggle>
        </PanelHeader>
        <PanelBody visible={toggled}>{children}</PanelBody>
      </Panel>
    );
  }
}

export default PanelComponent;
