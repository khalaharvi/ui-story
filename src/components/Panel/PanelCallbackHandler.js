import { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Helper wrapper for <xxxxFacet /> components wrapped in with <Panel /> component.
 * It is reponsible to update 'canRefine' state in <xxxxFacet />'s parent <Panel /> component.
 *
 * Usage:
 * - Export optionally <Panel />-wrapped <xxxxFacet /> component like so:
 *    <PanelCallbackHandler>
 *      <xxxxFacet />
 *    </PanelCallbackHandler>
 *
 * - Wrap <PanelCallbackHandler>-wrapped <xxxxFacet /> component with <Panel />
 *    <Panel>
 *      <xxxxFacet />
 *    </Panel>
 *
 * From: https://github.com/algolia/react-instantsearch/blob/v5.0.1/packages/react-instantsearch/src/components/PanelCallbackHandler.js
 */
class PanelCallbackHandler extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    canRefine: PropTypes.bool.isRequired
  };

  static contextTypes = {
    setCanRefine: PropTypes.func
  };

  componentWillMount() {
    if (this.context.setCanRefine) {
      this.context.setCanRefine(this.props.canRefine);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.context.setCanRefine && this.props.canRefine !== nextProps.canRefine) {
      this.context.setCanRefine(nextProps.canRefine);
    }
  }

  render() {
    return this.props.children;
  }
}

export default PanelCallbackHandler;
