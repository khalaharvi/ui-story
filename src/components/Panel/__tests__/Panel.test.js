import React from 'react';
import { shallow } from 'enzyme';

import Panel from '../Panel';

// Adapted from: https://github.com/algolia/react-instantsearch/blob/v5.0.1/packages/react-instantsearch/src/components/Panel.test.js
describe('<Panel />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Panel header="Panel Header">
        <div>This is the panel body.</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className', () => {
    const wrapper = shallow(
      <Panel header="Panel Header" className="custom-panel">
        <div>This is the panel body.</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with headerAction', () => {
    const wrapper = shallow(
      <Panel header="Panel Header" headerAction={<button>Click Me</button>}>
        <div>This is the panel body.</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly without refinement', () => {
    const wrapper = shallow(
      <Panel header="Panel Header">
        <div>This is the panel body.</div>
      </Panel>
    );

    wrapper.setState({ canRefine: false });

    expect(wrapper).toMatchSnapshot();
  });

  it('should expose context to his children', () => {
    const wrapper = shallow(
      <Panel header="Panel Header">
        <div>This is the panel body.</div>
      </Panel>
    );

    expect(wrapper.instance().getChildContext()).toEqual({
      setCanRefine: wrapper.instance().setCanRefine
    });
  });

  it('renders correctly when setCanRefine is called', () => {
    const wrapper = shallow(
      <Panel header="Panel Header">
        <div>This is the panel body.</div>
      </Panel>
    );

    expect(wrapper.state('canRefine')).toBe(true);
    expect(wrapper).toMatchSnapshot();

    // Simulate context call
    wrapper
      .instance()
      .getChildContext()
      .setCanRefine(false);

    wrapper.update();

    expect(wrapper.state('canRefine')).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });
});
