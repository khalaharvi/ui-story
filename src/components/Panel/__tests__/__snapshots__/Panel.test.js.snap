// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel /> renders correctly 1`] = `
<Styled(div)
  hidden={false}
>
  <Styled(div)
    onClick={[Function]}
  >
    <span>
      Panel Header
    </span>
    <Styled(button)
      aria-label="Toggle Panel Header panel"
      type="button"
    >
      <Plus
        rotate={false}
      />
    </Styled(button)>
  </Styled(div)>
  <Styled(div)
    visible={false}
  >
    <div>
      This is the panel body.
    </div>
  </Styled(div)>
</Styled(div)>
`;

exports[`<Panel /> renders correctly when setCanRefine is called 1`] = `
<Styled(div)
  hidden={false}
>
  <Styled(div)
    onClick={[Function]}
  >
    <span>
      Panel Header
    </span>
    <Styled(button)
      aria-label="Toggle Panel Header panel"
      type="button"
    >
      <Plus
        rotate={false}
      />
    </Styled(button)>
  </Styled(div)>
  <Styled(div)
    visible={false}
  >
    <div>
      This is the panel body.
    </div>
  </Styled(div)>
</Styled(div)>
`;

exports[`<Panel /> renders correctly when setCanRefine is called 2`] = `
<Styled(div)
  hidden={true}
>
  <Styled(div)
    onClick={[Function]}
  >
    <span>
      Panel Header
    </span>
    <Styled(button)
      aria-label="Toggle Panel Header panel"
      type="button"
    >
      <Plus
        rotate={false}
      />
    </Styled(button)>
  </Styled(div)>
  <Styled(div)
    visible={false}
  >
    <div>
      This is the panel body.
    </div>
  </Styled(div)>
</Styled(div)>
`;

exports[`<Panel /> renders correctly with className 1`] = `
<Styled(div)
  hidden={false}
>
  <Styled(div)
    onClick={[Function]}
  >
    <span>
      Panel Header
    </span>
    <Styled(button)
      aria-label="Toggle Panel Header panel"
      type="button"
    >
      <Plus
        rotate={false}
      />
    </Styled(button)>
  </Styled(div)>
  <Styled(div)
    visible={false}
  >
    <div>
      This is the panel body.
    </div>
  </Styled(div)>
</Styled(div)>
`;

exports[`<Panel /> renders correctly with headerAction 1`] = `
<Styled(div)
  hidden={false}
>
  <Styled(div)
    onClick={[Function]}
  >
    <span>
      Panel Header
    </span>
    <Styled(button)
      aria-label="Toggle Panel Header panel"
      type="button"
    >
      <Plus
        rotate={false}
      />
    </Styled(button)>
  </Styled(div)>
  <Styled(div)
    visible={false}
  >
    <div>
      This is the panel body.
    </div>
  </Styled(div)>
</Styled(div)>
`;

exports[`<Panel /> renders correctly without refinement 1`] = `
<Styled(div)
  hidden={true}
>
  <Styled(div)
    onClick={[Function]}
  >
    <span>
      Panel Header
    </span>
    <Styled(button)
      aria-label="Toggle Panel Header panel"
      type="button"
    >
      <Plus
        rotate={false}
      />
    </Styled(button)>
  </Styled(div)>
  <Styled(div)
    visible={false}
  >
    <div>
      This is the panel body.
    </div>
  </Styled(div)>
</Styled(div)>
`;
